File: 0014-longest-common-prefix/0014-longest-common-prefix.py
class Solution(object):
    def longestCommonPrefix(self, strs):
        """
        :type strs: List[str]
        :rtype: str
        """
        
        def is_prefix(length):

            prefix = strs[0][:length]
            return all([i.startswith(prefix) for i in strs])
        
        start, end = 0, len(min(strs))

        mid = (start+end)//2

        while start<=end:
            if is_prefix(mid):
                start = mid+1
            else:
                end = mid-1
            mid = (start+end)//2

        
        return strs[0][:mid]

            

File: 0014-longest-common-prefix/0014-longest-common-prefix.scala
object Solution {
    def longestCommonPrefix(strs: Array[String]): String = {

        def is_prefix(length:Int): Boolean = {

        var prefix = strs(0).substring(0,length)
        strs.forall(_.startsWith(prefix))

        }

        var start = 0
        var end = strs.map(_.length).min
        var mid = (start+end)/2

        while(start<=end){
        if(is_prefix(mid)){
            start = mid+1
        }else{
            end = mid-1
        }
        mid = (start+end)/2
        }

        strs(0).substring(0,mid)

  }
}